ledger/ledger.go

```diff
@@ -682,11 +691,37 @@ func (l *Ledger) GenesisProto() config.ConsensusParams {
 	return l.genesisProto
 }
 
+// GenesisProtoVersion returns the initial protocol version for this ledger.
+func (l *Ledger) GenesisProtoVersion() protocol.ConsensusVersion {
+	return l.genesisProtoVersion
+}
+
 // GenesisAccounts returns initial accounts for this ledger.
 func (l *Ledger) GenesisAccounts() map[basics.Address]basics.AccountData {
 	return l.genesisAccounts
 }
 
+// BlockHdrCached returns the block header if available.
+// Expected availability range is [Latest - MaxTxnLife, Latest]
+// allowing (MaxTxnLife + 1) = 1001 rounds back loopback.
+// The depth besides the MaxTxnLife is controlled by DeeperBlockHeaderHistory parameter
+// and currently set to 1.
+// Explanation:
+// Clients are expected to query blocks at rounds (txn.LastValid - (MaxTxnLife + 1)),
+// and because a txn is alive when the current round <= txn.LastValid
+// and valid if txn.LastValid - txn.FirstValid <= MaxTxnLife
+// the deepest lookup happens when txn.LastValid == current => txn.LastValid == Latest + 1
+// that gives Latest + 1 - (MaxTxnLife + 1) = Latest - MaxTxnLife as the first round to be accessible.
+func (l *Ledger) BlockHdrCached(rnd basics.Round) (hdr bookkeeping.BlockHeader, err error) {
+	l.trackerMu.RLock()
```

Comment:
I think this should just be `BlockHdr`, not `BlockHdrCached`.  Just like with other lookup routines, if you ask for something with the `rnd` too early, it can fail. But we don't call them all `lookupResourceCached` and so on.  Those fail if `rnd` if before the dbround, this one fails if asked before txtail length. But the caller probably thinks of them very similarly.

Reply:
Sure I tried this. See https://github.com/algorand/go-algorand/pull/3935#discussion_r866410261
TLDR: `replay` deadlocks.